### ISSUE TITLE
Document encoding gets lost on Project.AddDocument

### ISSUE THREAD

**Issue by raffaeler** [dotnet/roslyn#24045]
**Version Used**:  2.6.1
**Steps to Reproduce**:
When entirely generating the syntax nodes in-memory, the SyntaxGenerator does not allow specifying an encoding.
The workaround is to re-create the syntax tree by using CSharpSyntaxTree.Create with a valid encoding.
Problem 1: how bad is this perf hit?

But even after having a valid encoding in the SyntaxTree, the AddDocument does not keep the information (see the last assert in the code below).
Again, the workaround is to re-apply the encoding after the compilation has been created, but it looks definitely a weird behavor.

- Am I missing something?
- Is there a better way to enforce the encoding for handcrafted in-memory syntax trees?

```
// Encoding is needed to avoid CS8055: Cannot emit debug information for a source text without encoding
private void TestEncoding1()
{
    var workspace = new AdhocWorkspace();
    var generator = SyntaxGenerator.GetGenerator(workspace, LanguageNames.CSharp);
    var project = workspace.CurrentSolution.AddProject("test", "test", LanguageNames.CSharp);

    var root = GetCode(generator);
    // code generated on the fly does not have encoding
    Debug.Assert(root.SyntaxTree.Encoding == null);

    var encoded = CSharpSyntaxTree.Create(root as CSharpSyntaxNode,
        null, string.Empty, Encoding.UTF8).GetRoot();
    // encoding is now part of the syntax tree
    Debug.Assert(encoded.SyntaxTree.Encoding != null);

    var document = project.AddDocument("test", encoded);
    var documentTree = document.GetSyntaxTreeAsync().Result;
    // why does this assert fails?
    Debug.Assert(documentTree.Encoding != null);
}

private SyntaxNode GetCode(SyntaxGenerator generator)
{
    return generator.NamespaceDeclaration("Foo",
        generator.ClassDeclaration("FooClass", null,
            Accessibility.Public, DeclarationModifiers.Partial,
            null, null, null));
}
```

**Expected Behavior**:
1. I expect there a way to specify the default encoding in the SyntaxGenerator
2. I expect that AddDocument keep the encoding stored in the specified syntax tree
3. I want to avoid any possible perf hit that would come from generating the text source code and parsed back in a new syntax tree.

**Actual Behavior**:
The encoding is missing (see asserts in the code)

**Comment by sharwell** [comment]
If you use the `AddDocument` overload that takes a `SourceText` instead of a `SyntaxTree`, you can specify the encoding there. The underlying issue could be related to, or possibly even the same as, #4767.

**Comment by raffaeler** [comment]
Hi Sam, as I wrote this is not a viable solution because I am building the syntax nodes in memory and the primary goal is to totally avoid the strings.

**Comment by sharwell** [comment]
@raffaeler I was commenting more as a workaround until the bug gets fixed. 😄

**Comment by raffaeler** [comment]
Thanks. As a workaround I re-created the syntaxtree manually specifying the encoding:
```
foreach (var tree in compilation.SyntaxTrees)
{
    var encoded = CSharpSyntaxTree.Create(tree.GetRoot() as CSharpSyntaxNode,
        null, string.Empty, Encoding.UTF8).GetRoot();
    compilation = compilation.ReplaceSyntaxTree(tree, encoded.SyntaxTree);
}
```
But this sounds definitely as an overkill and a perf hit.

**Comment by CyrusNajmabadi** [comment]
> But this sounds definitely as an overkill and a perf hit.

What perf hit do you see here?   It actually seems pretty reasonable to me.

**Comment by raffaeler** [comment]
@CyrusNajmabadi could you please explain what do you mean?
I dont'have any source text. The syntax nodes were just created as syntax nodes, not parsed from any string.
Now why should I transform the syntax nodes to a string and then parsed again in a new syntax tree just to specify the encoding?
BTW the encoding is required from the Emit method in order to produce the pdbs (which are desiredable even in absence of the source text file.

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi could you please explain what do you mean?

I was just responding to your comment that the following was overkill/perf-hit:

```c#
foreach (var tree in compilation.SyntaxTrees)
{
    var encoded = CSharpSyntaxTree.Create(tree.GetRoot() as CSharpSyntaxNode,
        null, string.Empty, Encoding.UTF8).GetRoot();
    compilation = compilation.ReplaceSyntaxTree(tree, encoded.SyntaxTree);
}
```

It's not clear to me why this is overkill or a perf hit.  This should be very cheap to do (as it's not actually reparsing anything).  it just making a handful of objects with the right data in it.  So it's unclear to me why the above isn't a good enogh solution to the problem.

Thanks!

**Comment by raffaeler** [comment]
The reason why I wrote is an overkill is because all the Roslyn model classes are immutable. 
- The `CSharpSyntaxTree.Create` creates a clone of the tree with the new encoding
- The `ReplaceSyntaxTree` creates a clone of the compilation with the new syntax tree

Also, It makes more sense preserving the encoding in the AddDocument instead of updating the compilation object.

**Comment by sharwell** [comment]
@raffaeler Since the types are immutable, the clones are extremely shallow (and thus very cheap to create).

**Comment by raffaeler** [comment]
I agree and understand that it is not copying all the tree, but I believe you agree that it is an additional operation that could/should be avoided ... zero is better than something :)
To give some context, this is used during the application runtime, not a tool for developers.

**Comment by sharwell** [comment]
@raffaeler I agree that this should be fixed because the API shouldn't be modifying the syntax tree during the addition process. I'm not particularly concerned with the performance here at this time (i.e. I do not believe that working around this issue will result in an observable performance problem).

**Comment by CyrusNajmabadi** [comment]
> The CSharpSyntaxTree.Create creates a clone of the tree with the new encoding

Hrmm... all that shoudl be done is creating a new object that simply points at the underlying shared syntax node (the 'green node' in Roslyn parlance).  This is very cheap.

>  zero is better than something

Sure.  But that's a very different claim that this is "overkill and a perf hit." Roslyn does this sort of things under the covers all the time without issue.  It's how many (most?) of our features work :)

**Comment by raffaeler** [comment]
I did not measure at this time so I can't say if it really is a perf hit or not, but I am concerned about additional operations to do at runtime as code generation itself is (of course) already costly.
